### PR TITLE
Auth gateway 401 handler doesn't refresh the token multiple times

### PR DESCRIPTION
**Pull request by paulstatezny**
## Auth gateway 401 handler doesn't refresh the token multiple times
When the oAuth token expires the `HttpAuthGateway.handle401()` method will attempt to refresh the token EVEN IF the token is already in the middle of being refreshed.

### ACs
1. Only one token refresh will be attempted at a given time

### Tasks
1. When `handle401()` is called it will set a loading flag in local storage and trigger a refresh call
1. If the loading flag is already set then `handle401()` will simply register a listener to the `TOKEN_REFRESH_SUCCESS` event. The listener will reattempt the api call that triggered the `handle401()` call in the first place.
1. When the token refresh attempt is completed successfully the `TOKEN_REFRESH_SUCCESS ` event will be fired
1. If the refresh fails then fire `TOKEN_REFRESH_FAILURE`

### CODE REVIEW

**Comment by paulstatezny** [comment]
(Like that ^^^ Works perfectly, but I want to do some cleanup and testing.)

**Comment by paulstatezny** [comment]
So this is most the way there. (Tests not complete.)

But there's one problem: it avoids extra token refreshes *as long as* all of the bulk 401 responses are returned **before** the token refresh request is finished. So if any of them take extra time to return, another token refresh request is fired. It goes like this:

```
# This circumstance is already accounted for
1. A bunch of requests fired.
2. First one returns 401. Instantly a refresh token request is fired off.
3. All the other requests return 401s, and they know to wait because there is a refresh going on.
4. Refresh finishes; all remaining requests are re-executed.

# This circumstance is NOT accounted for
1. A bunch of requests fired.
2. First one returns 401. Instantly a refresh token request is fired off.
3. All the other requests EXCEPT ONE return 401s. Most of them know to wait because there is a refresh going on.
4. Refresh finishes, and all original requests that have returned 401s are re-executed.
5. The "laggard" request from the original batch comes back with a 401. There is *not* currently a refresh happening, so it doesn't know that the token is already updated.
```

I'm pretty sure in practice that circumstance 2 will rarely occur. However, it's kind of a nice to have, to optimize performance to the full extend possible in this area.

It wouldn't be that hard to address. We'd just need to grab the current token in `gateway::apiRequest` and pass it to `gateway.handleError` so that whenever 401 is received, we check if the token has changed. If it has, immediately re-execute the request. If it hasn't, we know that there's still an expired token and to go ahead and refresh, provided a refresh isn't already occurring.

**Comment by paulstatezny** [comment]
:star: Tested on a project. It works!!

Full disclosure: I did *not* test every aspect of this. Spent like 2 hours debugging why my tests weren't working, and *weird* behavior. Found out that due to the nature and architecture of this module, it's really hard to test the kind of behavior being added in this issue.

If we want to test this stuff, I would recommend splitting up the logic into different modules or somehow extracting it into different units that can be tested more easily.

**Comment by zpchavez** [comment]
:hand:

File: __tests__/http/auth-gateway-test.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+var HOSTNAME     = 'HOSTNAME';
+var ACCESS_TOKEN = 'ACCESS_TOKEN';
+
+describe('auth-gateway', function () {
+    var HttpGateway, AuthGateway, authGateway, https;
+
+    beforeEach(function () {
+        https = sinon.spy();
+
+        HttpGateway = proxyquire(
```
Comment (zpchavez) [inline __tests__/http/auth-gateway-test.js]: Proxyquire doesn't work with webpack. You've got to use https://github.com/jhnns/rewire-webpack

**Comment by zpchavez** [comment]
:+1:

**Pull request merged**